File: backend/src/main/java/com/training/cvmanagementbe/service/ApprovalService.java
package com.training.cvmanagementbe.service;

import com.training.cvmanagementbe.dto.request.approvals.CancelDraftRequest;
import com.training.cvmanagementbe.dto.request.approvals.ReassignRequest;
import com.training.cvmanagementbe.dto.request.approvals.RejectDraftRequest;
import com.training.cvmanagementbe.dto.request.approvals.ReplyCommentRequest;
import com.training.cvmanagementbe.dto.response.approvals.*;
import com.training.cvmanagementbe.dto.response.configs.PagedResponse;
import org.springframework.data.domain.Pageable;

import java.util.List;
import java.util.UUID;

/*
 * The two-level approval flow.
 * - This phase covers entering the flow - submitting, assigning and opening a draft to read it.
 * Approving, rejecting, inline comments, cancelling and reassigning are added by the following
 * parts of this stage and all of them keep the same contract: a draft transition is a
 * compare-and-set and only the assignee may act on an open assignment.
 */
public interface ApprovalService {

    /*
     * Submits the caller's own draft for approval.
     * - Resolves the level-1 reviewer and moves the draft to PENDING_TECH_LEAD, or
     * straight to PENDING_HR when the only eligible tech lead is the submitter.
     * From that moment the content is read-only for everyone including the owner
     */
    DraftSubmitResponse submit(UUID draftId);

    // The caller's own open assignments, most urgent first. Never anyone else's.
    PagedResponse<ApprovalQueueItem> getQueue(Pageable pageable);

    // Reviews cancelled while assigned to the caller, with the reason. Same scope as the queue.
    PagedResponse<CancelledReviewResponse> getCancelledReviews(Pageable pageable);

    // Opens a draft for review. Answers 403 to anyone but the current assignee.
    DraftReviewResponse openForReview(UUID draftId);

    // Approves at the current level; level 2 also publishes. CAS on (status, assignee)
    DraftApproveResponse approve(UUID draftId);

    // Rejects at the current level.
    DraftRejectResponse reject(UUID draftId, RejectDraftRequest request);

    // Re-submits a rejected draft for approval.
    DraftSubmitResponse resubmit(UUID draftId);

    // Replies to an inline comment.
    InlineCommentResponse replyToComment(UUID commentId, ReplyCommentRequest request);

    // ---------- Admin oversight ----------

    // Every draft under review, whoever holds it. Admin only; enforced at the controller.
    PagedResponse<PendingDraftResponse> getPendingDrafts(Pageable pageable);

    // Owner cancels DRAFT/REJECTED; Admin cancels any open draft, with a reason when it is pending.
    DraftCancelResponse cancel(UUID draftId, CancelDraftRequest request);

    // People who may take over the open assignment of this draft, the least loaded first.
    List<ReassignCandidateResponse> getReassignCandidates(UUID draftId);

    // Closes the open assignment as REASSIGNED and opens a new one with a fresh SLA.
    ReassignResponse reassign(UUID draftId, ReassignRequest request);
}

File: frontend/src/app/core/services/approval.service.ts
import { HttpClient, HttpParams } from "@angular/common/http";
import { inject, Injectable } from "@angular/core";
import { environment } from "../../../environments/environment";
import { ApprovalQueueItem, ApprovalQueueQuery, CancelDraftRequest, CancelledReviewItem, CancelledReviewQuery, DraftApproveResponse, DraftCancelResponse, DraftReviewResponse, DraftSubmitResponse, PendingDraftItem, PendingDraftQuery, ReassignCandidate, ReassignRequest, ReassignResponse } from "../dtos/approval.dto";
import { Observable } from "rxjs";
import { PagedResponse } from "../dtos/page.dto";
import { ApiEndpoint } from "../enums/api-endpoint.enum";
import { DraftRejectResponse, InlineCommentResponse, RejectDraftRequest, ReplyCommentRequest } from "../dtos/inline-comment.dto";

/*
 * The approval flow's client.
 *
 * No role check happens here or in the components that call it: the backend scopes the queue by
 * assignment and answers 403 on a draft that is not the caller's to review, so a guard in front of
 * these calls would only duplicate a rule it cannot enforce.
 */
@Injectable({ providedIn: 'root' })
export class ApprovalService {

    private readonly http = inject(HttpClient);

    private url(endpoint: string): string {
        return `${environment.apiBaseUrl}${endpoint}`;
    }

    getQueue(query: ApprovalQueueQuery): Observable<PagedResponse<ApprovalQueueItem>> {
        let params = new HttpParams()
            .set('page', query.page)
            .set('size', query.size);

        if (query.sortBy) {
            params = params.set('sortBy', query.sortBy);
        }
        if (query.direction) {
            params = params.set('direction', query.direction);
        }
        return this.http.get<PagedResponse<ApprovalQueueItem>>(
            this.url(ApiEndpoint.ApprovalQueue), { params }
        );
    }

    getCancelledReviews(query: CancelledReviewQuery): Observable<PagedResponse<CancelledReviewItem>> {
        const params = new HttpParams()
            .set('page', query.page)
            .set('size', query.size);

        return this.http.get<PagedResponse<CancelledReviewItem>>(
            this.url(`${ApiEndpoint.ApprovalQueue}/cancelled`), { params }
        );
    }

    // Locks the draft's content on success; the caller should navigate away rather than re-edit.
    submit(draftId: string): Observable<DraftSubmitResponse> {
        return this.http.post<DraftSubmitResponse>(
            this.url(`${ApiEndpoint.Approvals}/drafts/${draftId}/submit`), null
        );
    }

    resubmit(draftId: string): Observable<DraftSubmitResponse> {
        return this.http.post<DraftSubmitResponse>(
            this.url(`${ApiEndpoint.Approvals}/drafts/${draftId}/resubmit`), null
        );
    }

    openForReview(draftId: string): Observable<DraftReviewResponse> {
        return this.http.get<DraftReviewResponse>(
            this.url(`${ApiEndpoint.Approvals}/drafts/${draftId}/review`)
        );
    }

    approve(draftId: string): Observable<DraftApproveResponse> {
        return this.http.post<DraftApproveResponse>(
            this.url(`${ApiEndpoint.Approvals}/drafts/${draftId}/approve`), null
        );
    }

    reject(draftId: string, body: RejectDraftRequest): Observable<DraftRejectResponse> {
        return this.http.post<DraftRejectResponse>(
            this.url(`${ApiEndpoint.Approvals}/drafts/${draftId}/reject`), body
        );
    }

    replyToComment(commentId: string, body: ReplyCommentRequest): Observable<InlineCommentResponse> {
        return this.http.post<InlineCommentResponse>(
            this.url(`${ApiEndpoint.Approvals}/comments/${commentId}/reply`), body
        );
    }

    // ---------- Admin oversight ----------

    // Not scoped to an assignee, unlike getQueue: supervision means seeing other people's work.
    getPendingDrafts(query: PendingDraftQuery): Observable<PagedResponse<PendingDraftItem>> {
        let params = new HttpParams()
            .set('page', query.page)
            .set('size', query.size);

        if (query.sortBy) {
            params = params.set('sortBy', query.sortBy);
        }
        if (query.direction) {
            params = params.set('direction', query.direction);
        }
        return this.http.get<PagedResponse<PendingDraftItem>>(
            this.url(`${ApiEndpoint.Approvals}/drafts/pending`), { params}
        );
    }

    // The body is omitted by the owner; the server refuses an admin cancel without a reason.
    cancelDraft(draftId: string, body?: CancelDraftRequest): Observable<DraftCancelResponse> {
        return this.http.post<DraftCancelResponse>(
            this.url(`${ApiEndpoint.Approvals}/drafts/${draftId}/cancel`), body ?? {}
        );
    }

    getReassignCandidates(draftId: string): Observable<ReassignCandidate[]> {
        return this.http.get<ReassignCandidate[]>(
            this.url(`${ApiEndpoint.Approvals}/drafts/${draftId}/reassign-candidates`)
        );
    }

    reassign(draftId: string, body: ReassignRequest): Observable<ReassignResponse> {
        return this.http.post<ReassignResponse>(
            this.url(`${ApiEndpoint.Approvals}/drafts/${draftId}/reassign`), body
        );
    }
}